### PR TITLE
Introduce embree ray tracing

### PR DESCRIPTION
**Pull request by fedormatantsev**
Work in progress, but reviews would be of a great use at any time.
[UPDATE] I think it's ready to merge

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -165,6 +168,7 @@ if (CMAKE_SYSTEM_NAME STREQUAL "Windows")
 elseif (CMAKE_SYSTEM_NAME STREQUAL "Darwin")
     include (cmake/config/mac-clang.txt)
 elseif (CMAKE_SYSTEM_NAME MATCHES "Linux|FreeBSD")
+    add_definitions (-D_GLIBCXX_USE_CXX11_ABI=0)
```
Comment (fedormatantsev) [inline CMakeLists.txt]: todo: remove the line

File: src/appleseed/renderer/modeling/frame/frame.cpp [hunk 2]
```diff
@@ -222,7 +222,7 @@ void Frame::print_settings()
         "  crop window                   (%s, %s)-(%s, %s)\n"
         "  denoising mode                %s\n"
         "  render stamp                  %s\n"
-        "  save extra aovs               %s",
+        "  save extra aovs               %s\n",
```
Comment (fedormatantsev) [inline src/appleseed/renderer/modeling/frame/frame.cpp]: todo: remove \n

File: src/appleseed/renderer/modeling/object/triangle.h [hunk 3]
```diff
@@ -60,7 +60,7 @@ class Triangle
     foundation::uint32  m_n0, m_n1, m_n2;   // vertex normal indices
     foundation::uint32  m_a0, m_a1, m_a2;   // vertex attribute indices
     foundation::uint32  m_pa;               // primitive attribute index
-
+    foundation::uint32  m_align1, m_align2; // 16 byte alignment
```
Comment (fedormatantsev) [inline src/appleseed/renderer/modeling/object/triangle.h]: TODO: don't need those

File: CMakeLists.txt (hunk 1 above)
Comment (dictoon) [reply CMakeLists.txt]: Yes please!

File: CMakeLists.txt [hunk 4]
```diff
@@ -267,6 +271,13 @@ if (WITH_DISNEY_MATERIAL)
     endif ()
 endif ()
 
+if (WITH_EMBREE)
+    add_definitions (-DAPPLESEED_WITH_EMBREE)
+    if (USE_EXTERNAL_EMBREE)
+        find_package(Embree REQUIRED)
```
Comment (dictoon) [inline CMakeLists.txt]: Insert space after `find_package`.

File: CMakeLists.txt [hunk 5]
```diff
@@ -126,6 +126,9 @@ option (USE_STATIC_OCIO                     "Use static OpenColorIO libraries"
 option (USE_STATIC_OSL                      "Use static OpenShadingLanguage libraries"              ON)
 option (WARNINGS_AS_ERRORS                  "Treat compiler warnings as errors"                     ON)
 
+option (WITH_EMBREE                         "Use Embree intersection backend"                       ON)
+option (USE_EXTERNAL_EMBREE                 "Use external embree library"                           OFF)
```
Comment (dictoon) [inline CMakeLists.txt]: embree -> Embree

File: CMakeLists.txt [hunk 6]
```diff
@@ -342,6 +353,14 @@ if (WITH_DISNEY_MATERIAL)
     endif ()
 endif ()
 
+if (WITH_EMBREE)
+    if (NOT USE_EXTERNAL_EMBREE)
+        set (EMBREE_INCLUDE_DIRS ${APPLESEED_DEPS_STAGE_DIR}/include/)
```
Comment (dictoon) [inline CMakeLists.txt]: You're missing the Embree folder here.

File: cmake/config/linux-gcc.txt [hunk 7]
```diff
@@ -253,6 +253,14 @@ macro (link_against_seexpreditor target)
     target_link_libraries (${target} ${SEEXPREDITOR_LIBRARIES})
 endmacro ()
 
+macro (link_against_embree target)
+    if (USE_EXTERNAL_EMBREE)
+        target_link_libraries (${target} ${EMBREE_LIBRARIES})
+    else ()
+        #TODO
```
Comment (dictoon) [inline cmake/config/linux-gcc.txt]: Please do.

File: cmake/config/mac-clang.txt [hunk 8]
```diff
@@ -197,6 +197,14 @@ macro (link_against_seexpreditor target)
     target_link_libraries (${target} ${SEEXPREDITOR_LIBRARIES})
 endmacro ()
 
+macro (link_against_embree target)
+    if (USE_EXTERNAL_EMBREE)
+        target_link_libraries (${target} ${EMBREE_LIBRARIES})
+    else ()
+        #TODO
```
Comment (dictoon) [inline cmake/config/mac-clang.txt]: Please do.

File: cmake/config/win-vs.txt [hunk 9]
```diff
@@ -369,6 +369,14 @@ macro (link_against_seexpreditor target)
     endif ()
 endmacro ()
 
+macro (link_against_embree target)
+    if (USE_EXTERNAL_EMBREE)
+        target_link_libraries (${target} ${EMBREE_LIBRARIES})
+    else ()
+        #TODO
```
Comment (dictoon) [inline cmake/config/win-vs.txt]: I can help with this one.

File: cmake/modules/FindEmbree.cmake [hunk 10]
```diff
@@ -0,0 +1,64 @@
+
+#
+# This source file is part of appleseed.
+# Visit https://appleseedhq.net/ for additional information and resources.
+#
+# This software is released under the MIT license.
+#
+# Copyright (c) 2018 Fedor Matantsev, The appleseedhq Organization
```
Comment (dictoon) [inline cmake/modules/FindEmbree.cmake]: Did you actually write this file? Doesn't Embree provide one?

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 11]
```diff
@@ -1225,7 +1225,18 @@ namespace
 
             QVBoxLayout* layout = new QVBoxLayout();
             container()->setLayout(layout);
-
+            #ifdef APPLESEED_WITH_EMBREE
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: It's customary to omit indentation for preprocessor statements, and we follow that tradition (fix everywhere).

File: src/appleseed.studio/mainwindow/renderingsettingswindow.cpp [hunk 12]
```diff
@@ -1225,7 +1225,18 @@ namespace
...
+            create_system_use_embree_settings(layout);
+            create_direct_link("use_embree", "use_embree", "false");
+            if (config.get_parameters().strings().exist("use_embree"))
```
Comment (dictoon) [inline src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: Can we reduce this whole block to simply

    set_widget("use_embree", 
        config.get_parameters().get_optional<bool>("use_embree", false");

?

Comment (dictoon) [reply src/appleseed.studio/mainwindow/renderingsettingswindow.cpp]: In fact, since you've created a direct link just above, the `load_directly_linked_values()` call below should already take care of all this, so this code appears redundant to me.

File: src/appleseed/CMakeLists.txt [hunk 13]
```diff
@@ -989,6 +989,13 @@ set (renderer_kernel_intersection_sources
     renderer/kernel/intersection/triangletree.h
     renderer/kernel/intersection/trianglevertexinfo.h
 )
+
+if (WITH_EMBREE)
+    list (APPEND renderer_kernel_intersection_sources
+        renderer/kernel/intersection/embreescene.cpp
+        renderer/kernel/intersection/embreescene.h)
+endif()
```
Comment (dictoon) [inline src/appleseed/CMakeLists.txt]: Insert space after `endif` (in our CMakeLists.txt files all CMake keywords are followed by a space).

File: src/appleseed/foundation/math/fp.h [hunk 14]
```diff
@@ -366,7 +366,7 @@ inline float FP<float>::construct(
 
     const uint32 value =
           (sign     << 31)
-        | (exponent << 22)
+        | (exponent << 23)
```
Comment (dictoon) [inline src/appleseed/foundation/math/fp.h]: Please make a separate commit and PR for this.

File: src/appleseed/renderer/kernel/intersection/assemblytree.cpp [hunk 15]
```diff
@@ -173,6 +176,13 @@ void AssemblyTree::rebuild_assembly_tree()
         pretty_int(m_items.size()).c_str(),
         plural(m_items.size(), "assembly instance").c_str());
 
+    #ifdef APPLESEED_WITH_EMBREE
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/assemblytree.cpp]: You've put this code in the middle of the assembly tree construction, and in the middle of log messages related to that construction. Move this code somewhere else. I suggest the end of this method.

File: src/appleseed/renderer/kernel/intersection/assemblytree.cpp [hunk 16]
```diff
@@ -269,9 +279,12 @@ void AssemblyTree::update_tree_hierarchy()
...
         if (stored_version_it != m_assembly_versions.end())
         {
-            if (stored_version_it->second == current_version_id)
+            if ((stored_version_it->second == current_version_id)
+                #ifdef APPLESEED_WITH_EMBREE
+                && !m_dirty
+                #endif
+                )
             {
-                // The child trees of this assembly are up-to-date.
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/assemblytree.cpp]: Why did you remove this comment?

File: src/appleseed/renderer/kernel/intersection/assemblytree.cpp [hunk 17]
```diff
@@ -498,11 +524,52 @@ void AssemblyTree::create_curve_tree(const Assembly& assembly)
...
+                    assembly
+                )));
+        
+        scene = new Lazy<EmbreeScene>(move(embree_scene_factory));
+        m_embree_scene_repository.insert(hash, scene);
+    }
+
+    m_embree_scenes.insert(make_pair(assembly.get_uid(), scene));
+}
+#endif
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/assemblytree.cpp]: Insert a blank line above `#endif` for symmetry with `#ifdef`.

File: src/appleseed/renderer/kernel/intersection/assemblytree.h [hunk 18]
```diff
@@ -36,6 +36,7 @@
 #include "renderer/kernel/intersection/regiontree.h"
 #include "renderer/kernel/intersection/treerepository.h"
 #include "renderer/kernel/intersection/triangletree.h"
+#include "renderer/kernel/intersection/embreescene.h"
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/assemblytree.h]: Order alphabetically.

File: src/appleseed/renderer/modeling/scene/scene.h [hunk 19]
```diff
@@ -55,6 +55,9 @@ namespace renderer      { class Camera; }
 namespace renderer      { class OnFrameBeginRecorder; }
 namespace renderer      { class Project; }
 namespace renderer      { class SurfaceShader; }
+#ifdef APPLESEED_WITH_EMBREE
+namespace renderer      { class EmbreeDevice; }
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/scene.h]: Order alphabetically.

File: src/appleseed/renderer/modeling/scene/scene.cpp [hunk 20]
```diff
@@ -31,6 +31,7 @@
 #include "scene.h"
 
 // appleseed.renderer headers.
+#include "renderer/kernel/intersection/embreescene.h"
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/scene/scene.cpp]: Looks redundant. Remove?

File: src/appleseed/renderer/modeling/object/triangle.h (hunk 3 above)
Comment (dictoon) [reply src/appleseed/renderer/modeling/object/triangle.h]: Then please remove them.

File: src/appleseed/renderer/modeling/project/configuration.cpp [hunk 21]
```diff
@@ -181,6 +181,15 @@ Dictionary Configuration::get_metadata()
             .insert("label", "Render Threads")
             .insert("help", "Number of threads to use for rendering"));
 
+    #ifdef APPLESEED_WITH_EMBREE
+    metadata.insert(
+        "use_embree",
+        Dictionary()
+            .insert("type", "bool")
+            .insert("label", "Use Embree")
+            .insert("help", "Whether to use Embree ray tracing kernels or appleseed internal ones"));
```
Comment (dictoon) [inline src/appleseed/renderer/modeling/project/configuration.cpp]: *or appleseed internal ones* -> *instead of appleseed's internal ones*

File: src/appleseed/renderer/kernel/shading/shadingpoint.h [hunk 22]
```diff
@@ -281,6 +281,7 @@ class ShadingPoint
     friend class RendererServices;
     friend class ShadingPointBuilder;
     friend class TriangleLeafVisitor;
+    friend class EmbreeScene;
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/shadingpoint.h]: Order alphabetically.

File: src/appleseed/renderer/kernel/rendering/renderercomponents.cpp [hunk 23]
```diff
@@ -31,6 +31,7 @@
 
 // appleseed.renderer headers.
 #include "renderer/global/globallogger.h"
+#include "renderer/kernel/intersection/tracecontext.h"
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/renderercomponents.cpp]: Is this actually needed? If not, please revert.

File: src/appleseed/renderer/kernel/rendering/masterrenderer.cpp [hunk 24]
```diff
@@ -408,6 +409,18 @@ struct MasterRenderer::Impl
             return m_renderer_controller->get_status();
 
         // Build or update ray tracing acceleration structures.
+        #ifdef APPLESEED_WITH_EMBREE
+        if (m_params.strings().exist("use_embree"))
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/rendering/masterrenderer.cpp]: Simpler:

    m_project.get_trace_context().set_use_embree(
        m_params.get_optional<bool>("use_embree", false));

File: src/appleseed/renderer/kernel/shading/shadingpoint.cpp [hunk 25]
```diff
@@ -420,6 +420,7 @@ void ShadingPoint::refine_and_offset() const
         break;
 
       case PrimitiveProceduralSurface:
+        // TODO: do we need to refine & offset here as well?
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/shading/shadingpoint.cpp]: Good question!

File: src/appleseed/renderer/kernel/intersection/intersector.h [hunk 26]
```diff
@@ -35,6 +35,7 @@
 #include "renderer/kernel/intersection/intersectionsettings.h"
 #include "renderer/kernel/intersection/regiontree.h"
 #include "renderer/kernel/intersection/triangletree.h"
+#include "renderer/kernel/intersection/embreescene.h"
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/intersector.h]: Order alphabetically.

File: CMakeLists.txt (hunk 6 above)
Comment (fedormatantsev) [reply CMakeLists.txt]: I prefer using #include <embree3/...> instead. I find this notation much more informative.

consider
```cpp
#include <embree3/rtcore.h>
```
and
```cpp
#include <rtcore.h>
```

File: cmake/config/linux-gcc.txt (hunk 7 above)
Comment (fedormatantsev) [reply cmake/config/linux-gcc.txt]: I don't know what "external Embree" means in terms of linux build... Will be there once Windows build is there

File: cmake/modules/FindEmbree.cmake (hunk 10 above)
Comment (fedormatantsev) [reply cmake/modules/FindEmbree.cmake]: This one was authored by me. Embree doesn't provide one.

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp [hunk 27]
```diff
@@ -0,0 +1,518 @@
...
+#include "renderer/modeling/scene/containers.h"
+
+using namespace foundation;
+using namespace std;
+
+
+namespace
+{
+
+using namespace renderer;
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/embreescene.cpp]: Indent comment of anonymous namespaces.

The rule is: we always indent the content of namespaces. The exception is: we don't indent the content of _top level_ namespaces since that would amount to indenting the whole file which doesn't bring anything to the table.

Comment (dictoon) [reply src/appleseed/renderer/kernel/intersection/embreescene.cpp]: Move `using namespace renderer` clause at the translation unit level, next to the other ones.

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp [hunk 28]
```diff
@@ -0,0 +1,518 @@
...
+    size_t                  m_object_instance_idx;
+    uint32                  m_vis_flags;
+    size_t                  m_motion_steps_count;
+
+    RTCGeometryType         m_geometry_type;
+    RTCGeometry             m_geometry_handle;
+
+    ~GeometryData()
+    {
+        delete[] m_vertices;
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/embreescene.cpp]: You must initialize these pointers to `nullptr` if you want that constructor to always work.

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp [hunk 29]
```diff
@@ -0,0 +1,518 @@
...
+        delete[] m_primitives;
+        rtcReleaseGeometry(m_geometry_handle);
+    }
+};
+
+typedef map<unsigned int, shared_ptr<GeometryData>> GeometryDataContainer;
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/embreescene.cpp]: Do we really need a `shared_ptr<>` here? Wouldn't a `unique_ptr<>` be enough? The former imposes a non-trivial performance penalty.

Comment (dictoon) [reply src/appleseed/renderer/kernel/intersection/embreescene.cpp]: What's the key here? The type (`unsigned int`) is suspicious.

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp [hunk 30]
```diff
@@ -0,0 +1,518 @@
...
+typedef GeometryDataContainer::const_iterator       GeometryDataContainerIterator;
+
+inline void collect_triangle_data(
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/embreescene.cpp]: You can remove the `inline` keyword here, it's useless.

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp [hunk 31]
```diff
@@ -0,0 +1,518 @@
...
+    const ObjectInstance&   object_instance,
+    GeometryData&           geometry_data)
+{
+    assert(geometry_data.m_geometry_type == RTC_GEOMETRY_TYPE_TRIANGLE);
+
+    // Retrieve object space -> assembly space transform for the object instance.
+    const Transformd& transform = object_instance.get_transform();
+
+    // Retrieve the object.
+    Object& object = object_instance.get_object();
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/embreescene.cpp]: Make const.

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp [hunk 32]
```diff
@@ -0,0 +1,518 @@
...
+
+    // TODO: remove regions.
+    // Retrieve the region kit of the object.
+    const Access<RegionKit> region_kit(&object.get_region_kit());
+
+    assert(region_kit->size() == 1);
+    const size_t region_idx = 0;
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/embreescene.cpp]: For now we have regions so you'll need to iterate over them.

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp [hunk 33]
```diff
@@ -0,0 +1,518 @@
...
+    const size_t vertices_count = tess->m_vertices.size();
+    geometry_data.m_vertices_count = vertices_count;
+    geometry_data.m_vertices_stride = sizeof(GVector3);
+
+    // Allocate memory for the vertices. Keep one extra vertex for padding.
+    geometry_data.m_vertices = new GVector3[vertices_count * motion_steps_count + 1];
+
+    for (size_t i = 0; i < vertices_count; ++i)
+    {
+        const GVector3 vertex_os = tess->m_vertices[i];
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/embreescene.cpp]: Const ref.

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp [hunk 34]
```diff
@@ -0,0 +1,518 @@
...
+        geometry_data.m_vertices[i] = transform.point_to_parent(vertex_os);
+    }
+
+    // Retrieve assembly space vertices.
+    for (size_t m = 1; m < motion_steps_count; ++m)
+    {
+        for (size_t i = 0; i < vertices_count; ++i)
+        {
+            const GVector3 vertex_os = tess->get_vertex_pose(i, m);
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/embreescene.cpp]: Const ref.

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp [hunk 35]
```diff
@@ -0,0 +1,518 @@
...
+
+    for (size_t i = 0; i < primitives_count; ++i)
+    {
+        geometry_data.m_primitives[i * 3] = tess->m_primitives[i].m_v0;
+        geometry_data.m_primitives[i * 3 + 1] = tess->m_primitives[i].m_v1;
+        geometry_data.m_primitives[i * 3 + 2] = tess->m_primitives[i].m_v2;
+    }
+};
+
+inline void collect_curve_data(
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/embreescene.cpp]: Remove `inline`.

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp [hunk 36]
```diff
@@ -0,0 +1,518 @@
...
+// Returns minimal tnear needed to compensate double to float transition of ray fields.
+inline float get_min_tnear(
+    const float dir_x,
+    const float dir_y,
+    const float dir_z,
+    const float org_x,
+    const float org_y,
+    const float org_z)
+{
+    float max_dir = max(abs(dir_x), abs(dir_y));
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/embreescene.cpp]: `foundation/math/minmax.h` provides 3-arg variants that you can use here for clarity.

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp [hunk 37]
```diff
@@ -0,0 +1,518 @@
...
+    max_dir = max(max_dir, abs(dir_z));
+
+    uint32 max_exp = max(
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/embreescene.cpp]: And here.

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp [hunk 38]
```diff
@@ -0,0 +1,518 @@
...
+
+    default:
+        assert(!"Unsupported geometry type.");
+        break;
+    }
+}
+
+// Returns minimal tnear needed to compensate double to float transition of ray fields.
+inline float get_min_tnear(
+    const float dir_x,
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/embreescene.cpp]: Origin first, direction second. We always use this order. Following the same order here will avoid confusion and mistakes.

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp [hunk 39]
```diff
@@ -0,0 +1,518 @@
...
+        
+        break;
+
+    default:
+        assert(!"Unsupported geometry type.");
+        break;
+    }
+}
+
+// Returns minimal tnear needed to compensate double to float transition of ray fields.
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/embreescene.cpp]: It would be great if you could detail the algorithm implemented here. I don't really understand it.

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp [hunk 40]
```diff
@@ -0,0 +1,518 @@
...
+        2047 << (23 - 11));
+
+    return offset / max_dir;
+}
+
+void shading_ray_to_embree_ray(
+    const ShadingRay&   shading_ray,
+    RTCRay&             embree_ray)
+{
+    embree_ray.dir_x    = static_cast<float>(shading_ray.m_dir.x);
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/embreescene.cpp]: Org then dir.

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp [hunk 41]
```diff
@@ -0,0 +1,518 @@
...
+
+EmbreeScene::~EmbreeScene()
+{
+    rtcReleaseScene(m_impl->m_scene);
+    delete m_impl;
+}
+
+EmbreeSceneFactory::EmbreeSceneFactory(const EmbreeScene::Arguments& arguments)
+  : m_arguments(arguments)
+{};
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/embreescene.cpp]: Remove semicolon.

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp [hunk 42]
```diff
@@ -0,0 +1,518 @@
...
+
+unique_ptr<EmbreeScene> EmbreeSceneFactory::create()
+{
+    return unique_ptr<EmbreeScene>(new EmbreeScene(m_arguments));
+}
+
+} // namespace renderer
```
Comment (dictoon) [inline src/appleseed/renderer/kernel/intersection/embreescene.cpp]: Add end-of-line.

File: cmake/config/mac-clang.txt (hunk 8 above)
Comment (fedormatantsev) [reply cmake/config/mac-clang.txt]: Keep it here since I don't have mac at my disposal

File: src/appleseed/renderer/modeling/scene/scene.cpp (hunk 20 above)
Comment (fedormatantsev) [reply src/appleseed/renderer/modeling/scene/scene.cpp]: No, I need it here to return EmbreeDevice type

File: src/appleseed/renderer/kernel/intersection/embreescene.cpp (hunk 32 above)
Comment (fedormatantsev) [reply src/appleseed/renderer/kernel/intersection/embreescene.cpp]: MeshObject itself never produces regions...

File: src/appleseed/renderer/modeling/project/configuration.cpp (hunk 21 above)
Comment (dictoon) [reply src/appleseed/renderer/modeling/project/configuration.cpp]: Seems like you forgot to do this change?

File: src/appleseed/renderer/modeling/scene/scene.cpp (hunk 20 above)
Comment (dictoon) [reply src/appleseed/renderer/modeling/scene/scene.cpp]: I see.

Maybe it would make sense to split `embreescene.h` into `embreescene.h` and `embreedevice.h`.

**Comment by dictoon** [comment]
Superseded by https://github.com/appleseedhq/appleseed/pull/2154
